
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../assets/images/favicon.ico">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-8.3.9">
    
    
      
        <title>Algorithm description - iLund4u</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.1dff34a1.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_markdown_exec_pyodide.css">
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/material.css">
    
      <link rel="stylesheet" href="../../css/mkdocstrings.css">
    
      <link rel="stylesheet" href="https://unpkg.com/katex@0/dist/katex.min.css">
    
    <script>__md_scope=new URL("../..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  


  <script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-2MTBEQBNYC"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&gtag("event","search",{search_term:this.value})}),"undefined"!=typeof location$&&location$.subscribe(function(e){gtag("config","G-2MTBEQBNYC",{page_path:e.pathname})})})</script>
  <script async src="https://www.googletagmanager.com/gtag/js?id=G-2MTBEQBNYC"></script>


    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="ilund4u_blue" data-md-color-accent="ilund4u_yellow">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#algorithm-description" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="iLund4u" class="md-header__button md-logo" aria-label="iLund4u" data-md-component="logo">
      
  
  <?xml version="1.0" encoding="UTF-8"?>
<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" width="500" height="500" viewBox="0 0 500 500">
  <path d="M233.703576745615464,427.256416679763788c7.333369773565209,4.493880934460321,16.504100586736058,4.58153700476214,23.921944907303441.228507746331161,50.132626006159626-29.419434862202252,238.137964177803951-148.749554434098172,200.44862634459605-263.45746735486955-.038701212821252-.160953878998953-.09111622448836-.295231716696435-.143214562194771-.456303087769811-13.348463600363175-43.859509781060297-53.757954843338666-75.923165964325563-101.793264469624773-76.344433085516357-27.075997864574674-.23745506717114-51.89797185122552,9.610385093721561-70.942033621216979,26.067702736280808l-.01351461696504.013279632835292c-.013513900568796.013197943029809-.040343952648982.026360800184193-.053858569615841.039640433006753-1.48227180887443,1.124700009913795-25.064705234497524,19.201953708738074-37.35273767684339,39.104905121658703-10.847787872680783-19.168879117740289-29.646930734423222-35.221834687822593-35.175457092377655-39.740973548275178-18.842202906597777-17.244625280741275-43.839161852159123-27.127838666213393-71.303267889790732-27.36869742151066-48.048788440949465-.421385329678742-89.005043205677794,31.397142829636323-103.120685642969875,75.01579385540299-.054996971637593.1601321508806-.114976038024906-.630068064120678-.156494195023697-.46981770473667C-1.665428560296277,273.921462364258332,184.128852320982332,396.877080057964122,233.703576745615464,427.256416679763788Z" style="fill: none; stroke: #fff; stroke-miterlimit: 10; stroke-width: 43px;"/>
</svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            iLund4u
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Algorithm description
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="ilund4u_blue" data-md-color-accent="ilund4u_yellow"  aria-hidden="true"  type="radio" name="__palette" id="__palette_1">
          
        
      </form>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/art-egorov/ilund4u.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  <li class="md-tabs__item">
    <a href="../.." class="md-tabs__link">
      Home
    </a>
  </li>

      
        
  
  
    
  


  <li class="md-tabs__item">
    <a href="./" class="md-tabs__link md-tabs__link--active">
      Algorithm description
    </a>
  </li>

      
        
  
  


  <li class="md-tabs__item">
    <a href="../../ExampleDrivenGuide/cmd_guide/" class="md-tabs__link">
      Example-driven guide
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../Parameters/cmd_parameters/" class="md-tabs__link">
        Parameters
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../API/usage_examples/" class="md-tabs__link">
        API
      </a>
    </li>
  

      
        
  
  


  <li class="md-tabs__item">
    <a href="../../VersionLog/versions/" class="md-tabs__link">
      Version log
    </a>
  </li>

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="iLund4u" class="md-nav__button md-logo" aria-label="iLund4u" data-md-component="logo">
      
  
  <?xml version="1.0" encoding="UTF-8"?>
<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" width="500" height="500" viewBox="0 0 500 500">
  <path d="M233.703576745615464,427.256416679763788c7.333369773565209,4.493880934460321,16.504100586736058,4.58153700476214,23.921944907303441.228507746331161,50.132626006159626-29.419434862202252,238.137964177803951-148.749554434098172,200.44862634459605-263.45746735486955-.038701212821252-.160953878998953-.09111622448836-.295231716696435-.143214562194771-.456303087769811-13.348463600363175-43.859509781060297-53.757954843338666-75.923165964325563-101.793264469624773-76.344433085516357-27.075997864574674-.23745506717114-51.89797185122552,9.610385093721561-70.942033621216979,26.067702736280808l-.01351461696504.013279632835292c-.013513900568796.013197943029809-.040343952648982.026360800184193-.053858569615841.039640433006753-1.48227180887443,1.124700009913795-25.064705234497524,19.201953708738074-37.35273767684339,39.104905121658703-10.847787872680783-19.168879117740289-29.646930734423222-35.221834687822593-35.175457092377655-39.740973548275178-18.842202906597777-17.244625280741275-43.839161852159123-27.127838666213393-71.303267889790732-27.36869742151066-48.048788440949465-.421385329678742-89.005043205677794,31.397142829636323-103.120685642969875,75.01579385540299-.054996971637593.1601321508806-.114976038024906-.630068064120678-.156494195023697-.46981770473667C-1.665428560296277,273.921462364258332,184.128852320982332,396.877080057964122,233.703576745615464,427.256416679763788Z" style="fill: none; stroke: #fff; stroke-miterlimit: 10; stroke-width: 43px;"/>
</svg>

    </a>
    iLund4u
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/art-egorov/ilund4u.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Algorithm description
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Algorithm description
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#hotspot-annotation-mode" class="md-nav__link">
    Hotspot annotation mode
  </a>
  
    <nav class="md-nav" aria-label="Hotspot annotation mode">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#defining-groups-of-protein-homologues-and-proteome-filtering" class="md-nav__link">
    Defining groups of protein homologues and proteome filtering
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#proteome-network-construction" class="md-nav__link">
    Proteome network construction
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#proteome-community-detection" class="md-nav__link">
    Proteome community detection
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#defining-protein-group-classes" class="md-nav__link">
    Defining protein group classes
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation-of-variable-islands" class="md-nav__link">
    Annotation of variable islands
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#island-network-construction-and-hotspot-annotation" class="md-nav__link">
    Island network construction and hotspot annotation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hotspot-merging" class="md-nav__link">
    Hotspot merging
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#additional-functional-annotation" class="md-nav__link">
    Additional functional annotation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#lovis4u-visualisation" class="md-nav__link">
    LoVis4u visualisation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ilund4u-database" class="md-nav__link">
    iLund4u database
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#protein-annotation-mode" class="md-nav__link">
    Protein annotation mode
  </a>
  
    <nav class="md-nav" aria-label="Protein annotation mode">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#homology-search" class="md-nav__link">
    Homology search
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hotspots-retrieving" class="md-nav__link">
    Hotspots retrieving
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#proteome-annotation-mode" class="md-nav__link">
    Proteome annotation mode
  </a>
  
    <nav class="md-nav" aria-label="Proteome annotation mode">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#homology-search_1" class="md-nav__link">
    Homology search
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#defining-proteome-community" class="md-nav__link">
    Defining proteome community
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation-of-variable-islands-and-hotspot-mapping" class="md-nav__link">
    Annotation of variable islands and hotspot mapping
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../ExampleDrivenGuide/cmd_guide/" class="md-nav__link">
        Example-driven guide
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Parameters
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Parameters" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Parameters
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../Parameters/cmd_parameters/" class="md-nav__link">
        Command-line parameters
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../Parameters/config_parameters/" class="md-nav__link">
        Configuration file parameters
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          API
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="API" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          API
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../API/usage_examples/" class="md-nav__link">
        Usage examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../API/package/" class="md-nav__link">
        Library
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../VersionLog/versions/" class="md-nav__link">
        Version log
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="algorithm-description">Algorithm description</h1>
<p>Here we describe in details each step of the iLund4u workflow, which schematically is shown on the Figure 1. For description of generated files, visualisation, and frequently used parameters, see the <a href="https://art-egorov.github.io/ilund4u/ExampleDrivenGuide/cmd_guide/">Example-driven guide page</a>. </p>
<figure style="align:center; margin-top: 0; margin-bottom: 0;">
    <a class="glightbox" href="../../img/ilundu4_pipeline.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img alt="pipeline" loading="lazy" src="../../img/ilundu4_pipeline.png" width="100%" /></a><br />
    <figcaption style="margin-top: 0; margin-bottom: 0;">Figure 1. iLund4u workflow.</figcaption>
</figure>
<h2 id="hotspot-annotation-mode">Hotspot annotation mode</h2>
<h3 id="defining-groups-of-protein-homologues-and-proteome-filtering">Defining groups of protein homologues and proteome filtering</h3>
<p>The first step is reading input data in gff format with <a href="https://github.com/chapmanb/bcbb/tree/master/gff">BCBio-gff</a> library. While reading the files we store all encoded protein sequences in a fasta file which then will be used as query for MMseqs clustering. In the iLund4u objects we do not store sequences, instead of that iLund4u objects hold links to the corresponding files (a gff file for proteomes and the fasta file with amino acid sequences) which helps to avoid large memory usage.<br />
During this step we have the first filtering based on the proteome size - we do not take for analysis proteomes with number of proteins <span class="arithmatex">\(N_{CDSs} &lt; 15\)</span>. The number of minimal proteome size can be changed with <code>-mps, --min-proteome-size &lt;int&gt;</code> parameter or config parameter <code>min_proteome_size</code>. </p>
<p>After gff files are read we use <a href="https://github.com/soedinglab/mmseqs2">MMseqs2</a> clustering pipeline with all encoded proteins as query. All parameters of clustering can be adjusted using configuration file (see <a href="https://art-egorov.github.io/ilund4u/Parameters/cmd_parameters/">Parameters</a> section). By default, we use the following parameters: <code>--cluster mode 0 --cov-mode 0 --min-seq-id 0.25 -c 0.8 -s 7</code> (see <a href="https://mmseqs.com/latest/userguide.pdf">mmseqs documentation</a> for detailed description).</p>
<p>We then postulate that each cluster of proteins is a group of homologues. For each group we set identifier (id) attribute as the id of the representative protein for the cluster. </p>
<p>While processing the MMseqs results we iterate over all proteomes and all encoded proteins of each proteome setting each CDS attribute <em>«group»</em> as identifier of corresponding to this CDS protein cluster. Additionally, on this step we perform deduplication - if over iteration we have a proteome which set of protein groups is found on already processed proteomes - then this proteome will be excluded from the analysis. That is, we do not keep for subsequent steps proteomes with identical proteome composition in terms of groups of homologues. We have another filtering process at this step: we exclude for the analysis proteomes which proteome composition uniqueness is less than set cutoff [default: 0.7] (can be changed in config file with <em>proteome_uniqueness_cutoff</em> parameter). Proteome uniqueness is defined as following: <span class="arithmatex">\(pu_{i} = {|G_i|}/{N_i}\)</span>, where <span class="arithmatex">\(|G_i|\)</span> is length (number of unique elements) of the set of protein groups of <span class="arithmatex">\(i^{th}\)</span> proteome, and <span class="arithmatex">\(N_i\)</span> - number of CDSs of <span class="arithmatex">\(i^{th}\)</span> proteome. In other word, we want to exclude from the analysis proteomes that contain a lot of duplicated proteins which is often a sign of an incorrect assembly if the defined uniqueness is low.</p>
<h3 id="proteome-network-construction">Proteome network construction</h3>
<p>The second major step is building the network of proteomes. The idea is to construct a weighted graph where each node, say <span class="arithmatex">\(n_i\)</span>, represents the corresponding <span class="arithmatex">\(i^{th}\)</span> proteome. Each graph edge, say <span class="arithmatex">\(e_{ij}\)</span>, connecting <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> proteomes has weight <span class="arithmatex">\(w_{ij}:\)</span>
$$
w_{ij} = (\frac{|G_i \cap G_j|}{|G_i|} + \frac{|G_i \cap G_j|}{|G_j|})\cdot\frac{1}{2} = \frac{|G_i \cap G_j|\cdot(|G_i|+|G_j|)}{2|G_i||G_j|}; \,\, w_{ij}\in[0,1]; \, \, w_{ij} = w_{ji} \,\, \forall\,i,j
$$ 
where <span class="arithmatex">\(|G_i|\)</span>, <span class="arithmatex">\(|G_j|\)</span> are lengths (number of unique elements) of the sets of protein groups of <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> proteomes, respectively. <span class="arithmatex">\(|G_i \cap G_j|\)</span> is the number of shared (overlapped) protein groups between <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> proteomes.<br />
The meaning of <span class="arithmatex">\(w_{ij}\)</span> is average value of two fractions: the first <span class="arithmatex">\(|G_i \cap G_j|/|G_i|\)</span> is the number of shared homologues normalised to the size of <span class="arithmatex">\(i^{th}\)</span> proteome, and the second <span class="arithmatex">\(|G_i \cap G_j|/|G_j|\)</span> - the same number of shared homologues with normalisation to the size of <span class="arithmatex">\(j^{th}\)</span> proteome. Average of two normalisation gives needed symmetric property with respect to its indexes. That is, the defined <span class="arithmatex">\(w_{ij}\)</span> is a metric of similarity of two genomes in terms of their proteome composition.</p>
<p>While building the graph we write only edges, which weight passes the set cutoff, by default if <span class="arithmatex">\(w_{ij}\geq 0.7\)</span> (the cutoff can be changed with config parameter <em>proteome_similarity_cutoff</em>). This not only reduces memory usage for a large set of input data but also allows us to skip calculation of those <span class="arithmatex">\(w_{ij}\)</span> if for corresponding <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> proteomes we have <span class="arithmatex">\(|G_j| &gt; |G_i|/0.7\)</span>. In other word, if we cannot reach weight 0.7 due to the proteome size difference we do not start its calculation.<br />
To implement this idea in the algorithm we proceed as follows: we sort all proteomes by length of protein group set and then for each unique length value <span class="arithmatex">\(l\)</span> we calculate upper length cutoff <span class="arithmatex">\(L^u(l)\)</span>: <span class="arithmatex">\(L^u(l)=l/0.7\)</span>. That allows us to limit the set of potential network connection indexes <span class="arithmatex">\(c_i\)</span> for <span class="arithmatex">\(i^{th}\)</span> proteome, which we need to calculate, from <span class="arithmatex">\(i &lt; c_i\)</span> to <span class="arithmatex">\(i &lt; c_i &lt; L^u(|G_i|)\)</span>.  </p>
<p>Below we show the python-like pseudocode of the proteome network construction function. </p>
<details class="card_heart_cog" open="open">
<summary><strong>Network construction pseudocode</strong></summary>
<div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-0-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-0-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-0-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-0-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-0-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-0-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-0-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-0-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-0-10">10</a></span>
<span class="normal"><a href="#__codelineno-0-11">11</a></span>
<span class="normal"><a href="#__codelineno-0-12">12</a></span>
<span class="normal"><a href="#__codelineno-0-13">13</a></span>
<span class="normal"><a href="#__codelineno-0-14">14</a></span>
<span class="normal"><a href="#__codelineno-0-15">15</a></span>
<span class="normal"><a href="#__codelineno-0-16">16</a></span>
<span class="normal"><a href="#__codelineno-0-17">17</a></span>
<span class="normal"><a href="#__codelineno-0-18">18</a></span>
<span class="normal"><a href="#__codelineno-0-19">19</a></span>
<span class="normal"><a href="#__codelineno-0-20">20</a></span>
<span class="normal"><a href="#__codelineno-0-21">21</a></span>
<span class="normal"><a href="#__codelineno-0-22">22</a></span>
<span class="normal"><a href="#__codelineno-0-23">23</a></span>
<span class="normal"><a href="#__codelineno-0-24">24</a></span>
<span class="normal"><a href="#__codelineno-0-25">25</a></span>
<span class="normal"><a href="#__codelineno-0-26">26</a></span>
<span class="normal"><a href="#__codelineno-0-27">27</a></span>
<span class="normal"><a href="#__codelineno-0-28">28</a></span>
<span class="normal"><a href="#__codelineno-0-29">29</a></span>
<span class="normal"><a href="#__codelineno-0-30">30</a></span>
<span class="normal"><a href="#__codelineno-0-31">31</a></span>
<span class="normal"><a href="#__codelineno-0-32">32</a></span>
<span class="normal"><a href="#__codelineno-0-33">33</a></span>
<span class="normal"><a href="#__codelineno-0-34">34</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1"></a><span class="c1"># Initialise the dictionary with the following structure: key - protein group of homologues (str),</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2"></a><span class="c1"># value - list of proteomes where it is encoded (as collections.deque object) </span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3"></a><span class="n">protein_group_to_proteome_index</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4"></a>    <span class="s2">&quot;protein_group_1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">ind1_1</span><span class="p">,</span> <span class="n">ind1_2</span><span class="p">,</span> <span class="n">ind1_3</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5"></a>    <span class="s2">&quot;protein_group_2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">ind2_1</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="o">...</span> <span class="p">}</span>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6"></a><span class="c1"># Initialise the dictionary of upper index cutoffs:</span>
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7"></a><span class="c1"># key - size of protein group set (int), value - upper proteome index cutoff (int)</span>
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8"></a><span class="n">upper_index_cutoff</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9"></a>    <span class="s2">&quot;protein_group_set_size_1&quot;</span><span class="p">:</span> <span class="n">upper_proteome_index_cutoff1</span><span class="p">,</span> <span class="o">...</span> <span class="p">}</span>
</span><span id="__span-0-10"><a id="__codelineno-0-10" name="__codelineno-0-10"></a><span class="c1"># Create a graph object to represent the proteome network</span>
</span><span id="__span-0-11"><a id="__codelineno-0-11" name="__codelineno-0-11"></a><span class="n">proteome_network</span> <span class="o">=</span> <span class="n">GraphObject</span><span class="p">()</span>
</span><span id="__span-0-12"><a id="__codelineno-0-12" name="__codelineno-0-12"></a><span class="c1"># Iterate over each proteome index</span>
</span><span id="__span-0-13"><a id="__codelineno-0-13" name="__codelineno-0-13"></a><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indexes_of_proteomes</span><span class="p">:</span>
</span><span id="__span-0-14"><a id="__codelineno-0-14" name="__codelineno-0-14"></a>    <span class="c1"># Initialise a dictionary to count number of shared homologues</span>
</span><span id="__span-0-15"><a id="__codelineno-0-15" name="__codelineno-0-15"></a>    <span class="n">counts_i</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
</span><span id="__span-0-16"><a id="__codelineno-0-16" name="__codelineno-0-16"></a>    <span class="c1"># Iterate over each protein group in the ith proteome</span>
</span><span id="__span-0-17"><a id="__codelineno-0-17" name="__codelineno-0-17"></a>    <span class="k">for</span> <span class="n">protein_group</span> <span class="ow">in</span> <span class="n">ith_proteome_protein_groups</span><span class="p">:</span>
</span><span id="__span-0-18"><a id="__codelineno-0-18" name="__codelineno-0-18"></a>       <span class="c1"># Iterate over each proteome index associated with the current protein group</span>
</span><span id="__span-0-19"><a id="__codelineno-0-19" name="__codelineno-0-19"></a>        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">protein_group_to_proteome_index</span><span class="p">[</span><span class="n">protein_group</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">():</span>
</span><span id="__span-0-20"><a id="__codelineno-0-20" name="__codelineno-0-20"></a>            <span class="c1"># Retrieve the upper limit for the current proteome size</span>
</span><span id="__span-0-21"><a id="__codelineno-0-21" name="__codelineno-0-21"></a>            <span class="n">upper_limit</span> <span class="o">=</span> <span class="n">upper_index_cutoff</span><span class="p">[</span><span class="n">ith_proteome_size</span><span class="p">]</span>
</span><span id="__span-0-22"><a id="__codelineno-0-22" name="__codelineno-0-22"></a>            <span class="c1"># Check if the current index j is within the acceptable range</span>
</span><span id="__span-0-23"><a id="__codelineno-0-23" name="__codelineno-0-23"></a>            <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">upper_limit</span><span class="p">:</span>
</span><span id="__span-0-24"><a id="__codelineno-0-24" name="__codelineno-0-24"></a>                <span class="n">counts_i</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
</span><span id="__span-0-25"><a id="__codelineno-0-25" name="__codelineno-0-25"></a>            <span class="k">elif</span> <span class="n">j</span> <span class="o">&lt;=</span> <span class="n">i</span><span class="p">:</span>
</span><span id="__span-0-26"><a id="__codelineno-0-26" name="__codelineno-0-26"></a>                <span class="c1"># Remove j from the upper index cutoff if it is less than or equal to i</span>
</span><span id="__span-0-27"><a id="__codelineno-0-27" name="__codelineno-0-27"></a>                <span class="n">upper_index_cutoff</span><span class="p">[</span><span class="n">ith_proteome_size</span><span class="p">]</span><span class="o">.</span><span class="n">popleft</span><span class="p">()</span>
</span><span id="__span-0-28"><a id="__codelineno-0-28" name="__codelineno-0-28"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-29"><a id="__codelineno-0-29" name="__codelineno-0-29"></a>                <span class="c1"># Exit the loop if j exceeds the upper limit</span>
</span><span id="__span-0-30"><a id="__codelineno-0-30" name="__codelineno-0-30"></a>                <span class="k">break</span>    
</span><span id="__span-0-31"><a id="__codelineno-0-31" name="__codelineno-0-31"></a>    <span class="c1"># Normalise the counts to get the weights</span>
</span><span id="__span-0-32"><a id="__codelineno-0-32" name="__codelineno-0-32"></a>    <span class="n">weights_i</span> <span class="o">=</span> <span class="n">normalise</span><span class="p">(</span><span class="n">counts_i</span><span class="p">)</span>
</span><span id="__span-0-33"><a id="__codelineno-0-33" name="__codelineno-0-33"></a>    <span class="c1"># Update the proteome network with new weights</span>
</span><span id="__span-0-34"><a id="__codelineno-0-34" name="__codelineno-0-34"></a>    <span class="n">proteome_network</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">weights_i</span><span class="p">)</span>      
</span></code></pre></div></td></tr></table></div>
</details>
<p>The complexity of the network construction step is <span class="arithmatex">\(O(N_p\cdot\bar{|G|}\cdot\bar{P})\)</span>, where <span class="arithmatex">\(N_p\)</span> is the number of proteomes, <span class="arithmatex">\(\bar{|G|}\)</span> is the average proteome size, and <span class="arithmatex">\(\bar{P}\)</span> is the average size of protein group of homologues (in how many proteomes its members are encoded). In general, it equals to the sum of shared homologues between proteomes that could be connected. </p>
<p>Based on our tests we have the following list of elapsed time for the network building function:<br />
i) 70 seconds for 20K Enterobacteria phage proteomes on M1 MacBook Pro laptop;<br />
ii) 500 seconds (8.4 minutes) for 222K plasmid proteomes on AMD 7413 (2.65 Ghz) computing node;  <br />
iii) 3610 seconds (60 minutes) for 563K phage proteomes on AMD 7413 (2.65 Ghz) computing node.<br />
Relatively big difference between phage and plasmid proteome processing is due to the much higher density of phage network (22M vs 860K written connections). </p>
<h3 id="proteome-community-detection">Proteome community detection</h3>
<p>When the network is constructed the next step is to define communities of proteomes. For this task we use the Leiden algorithm (<em><a href="https://doi.org/10.1038/s41598-019-41695-z">Traag Vincent A et al, Scientific reports (2019)</a></em>) of <a href="https://leidenalg.readthedocs.io/en/stable/index.html">leidenalg python library</a> implementation with the quality function based on the Constant Potts Model (CPM): <span class="arithmatex">\(Q =  \sum_{c}{e_c - \gamma \binom{n_c}{2}}\)</span> where <span class="arithmatex">\(e_c\)</span> is the weighted number of edges inside community <span class="arithmatex">\(c\)</span>, <span class="arithmatex">\(n_c\)</span> is the number of nodes in community <span class="arithmatex">\(c\)</span>, and <span class="arithmatex">\(\gamma\)</span> is the resolution parameter, which can be interpreted as inner and outer edge density threshold (see <em><a href="https://doi.org/10.1103/PhysRevE.84.016114">Traag Vincent A. et al Physical Review E—Statistical, Nonlinear, and Soft Matter Physics 2011</a></em>). By default the resolution parameter is set as: <span class="arithmatex">\(\gamma = 0.5\)</span> and can be changed with <em>leiden_resolution_parameter_p</em> config argument.</p>
<p>We postulate that each proteome community represents a group of related proteomes.</p>
<h3 id="defining-protein-group-classes">Defining protein group classes</h3>
<p>Then within each proteome community we can find protein groups of homologues which members are encoded in the majority of proteomes that we call <em>«conserved»</em> (also named as «core» in pan-genome analysis). If a protein group is found in a relatively small fraction of proteomes then we call it <em>«variable»</em> (named as «cloud» in pan-genome analysis), and, if a protein group presence is in the middle - we say that it is a representative of <em>«intermediate»</em> class («shell» in pan-genome analysis).</p>
<figure style="float: right; margin: 0; padding: 0;">
<a class="glightbox" href="../../img/protein_groups_fig.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img align="right" alt="f2" loading="lazy" margin="0" padding="0" src="../../img/protein_groups_fig.png" width="300" /></a><br />
<figcaption style="width: 100; margin: 0; padding: 0;">Figure 2</figcaption>
</figure>
<p>That is, for each protein group <span class="arithmatex">\(g_i\)</span> within a proteome community <span class="arithmatex">\(c\)</span> that consists of <span class="arithmatex">\(N_c\)</span> proteomes we define protein group presence fraction as <span class="arithmatex">\(f_i =  N_c^{g_i}/N_c\)</span>, where <span class="arithmatex">\(N_c^{g_i}\)</span> - is the number of proteomes of <span class="arithmatex">\(c\)</span> community that contain <span class="arithmatex">\(g_i\)</span> protein group. if <span class="arithmatex">\(f_i &lt; 0.25\)</span> <span class="arithmatex">\(g_i\)</span> group class is set as <em>«variable»</em>, if $ 0.25 \leq f_i \leq 0.75$ then <span class="arithmatex">\(g_i\)</span> is <em>«intermediate»</em>, and, if $ f_i &gt; 0.75 $ we say that <span class="arithmatex">\(g_i\)</span> is <em>«conserved»</em>. </p>
<p>In defined communities of proteome the number of «conserved» protein groups <span class="arithmatex">\(G_c^{con}\)</span> of any community, say community <span class="arithmatex">\(c\)</span>, will be proportional to the average proteome size <span class="arithmatex">\(\overline{|G|}_c\)</span> of community members, while the number of «variable» protein groups <span class="arithmatex">\(G_c^{var}\)</span> will be proportional to the <span class="arithmatex">\(N_c\)</span> - community size (number of proteomes). That is, <span class="arithmatex">\(G_c^{con} \propto \overline{|G|}_c\)</span> and <span class="arithmatex">\(G_c^{var} \propto N_c\)</span>.  </p>
<p>See the trivial example illustration on Figure 2 with the distribution of protein classes for a proteome community with all proteomes having <span class="arithmatex">\(K\)</span> proteins, <span class="arithmatex">\(K-1\)</span> being a member of «conserved» protein class with <span class="arithmatex">\(f = 1\)</span> and <span class="arithmatex">\(1\)</span> unique protein class per each proteome with <span class="arithmatex">\(f = 1/N\)</span>. </p>
<h3 id="annotation-of-variable-islands">Annotation of variable islands</h3>
<p>After we assigned protein group class to each protein within each proteome community, we can now define <em>«variable island»</em> objects. We say that a <em>variable island</em> is a region containing a set of adjacent and non-conserved proteins with at least one protein being «variable». The minimal size of island is one protein (and it should be «variable») while no upper limit is set. In other words, a <em>«variable island»</em> is a locus of «variable» protein(s) without internal interruption by «conserved» ORFs but possibly containing «intermediate» proteins. </p>
<p>Here it is important to note that the annotation process highly depends on the "circularity" property of a query genome. If a genome is circular then the first and the last genes will be consider as neighbours. By default we consider each genome as circular, which can be changed with <code>-ncg, --non-circular-genomes</code> parameter. Additionally, circularity property of each genome can be set independently with <code>-gct, --genome-circularity-table &lt;path&gt;</code> (see <a href="https://art-egorov.github.io/ilund4u/Parameters/cmd_parameters/">Parameters sections</a> for details). </p>
<p>The most important island attribute that defines all subsequent steps is its conserved neighbourhood. When a variable island region is defined we iterate over its flanking proteins on both sides. Over iteration we add each «conserved» protein to the attribute of conserved neighbours (either left or right). The iteration on each side stops either when five neighbours (adjustable with <em>neighbours_one_side_max_size</em> config parameter) are collected on this side or when we passed over eighth neighbour proteins distance (<em>neighbours_max_distance</em> config parameter). The minimal number of collected «conserved» neighbours is set by default as five (<em>neighbours_min_size</em> config parameter) and if an island does not reach the total number of five «conserved» neighbours it will not be taken for subsequent steps. The limit on distance from the island is set to avoid cases where an island have only few «conserved» proteins in flanking genes and the whole locus around consists of variable proteins, which will not allow us to analyse it as a hotspot member. </p>
<h3 id="island-network-construction-and-hotspot-annotation">Island network construction and hotspot annotation</h3>
<p>We postulate that a <em>hotspot</em> is a set of <em>variable islands</em> that have similar «conserved» flanking genes. From this definition we can consider each <em>variable island</em> as a locus characterised by its «conserved» neighbours and build a network of islands where each node, say, <span class="arithmatex">\(n_i^c\)</span>, represents the corresponding <span class="arithmatex">\(i^{th}\)</span> <em>variable island</em> in proteome community <span class="arithmatex">\(c\)</span> and each edge, say <span class="arithmatex">\(e_{ij}^c\)</span>, connecting <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> <em>variable islands</em> has weight <span class="arithmatex">\(w_{ij}^c\)</span>:<br />
$$
w_{ij}^{c} = \frac{|G_{ic} \cap G_{jc}|\cdot(|G_{ic}|+|G_{jc}|)}{2|G_{ic}||G_{jc}|}
$$
where <span class="arithmatex">\(|G_{ic}|\)</span>, <span class="arithmatex">\(|G_{jc}|\)</span> are lengths (number of unique elements) of the sets of protein groups of <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> island «conserved» neighbours, respectively. <span class="arithmatex">\(|G_{ic} \cap G_{jc}|\)</span> is the number of shared (overlapped) protein groups between <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> island «conserved» neighbours. Here the index <span class="arithmatex">\(c\)</span> representing the proteome community indicates that each community of proteome is treated independently and each community has its own network of variable islands. </p>
<p>As you see from the definition the network is identical to the network of proteomes if we consider each island as a new proteome with its conserved flanking genes (without cargo proteins) as its proteins. That allows us to apply the same algorithm of network construction and community detection as already was described in the section of proteome network building. We only have small differences in parameters: instead of minimal weight <span class="arithmatex">\(w_{ij}\geq 0.7\)</span> in the proteome network construction we have the threshold  <span class="arithmatex">\(w_{ij}^c\geq 0.65\)</span> (<em>island_neighbours_similarity_cutoff</em> config parameter) and the Leiden CPM resolution parameter <span class="arithmatex">\(\gamma\)</span> is now set as: <span class="arithmatex">\(\gamma = 0.55\)</span> (<em>leiden_resolution_parameter_i</em> config parameter) instead <span class="arithmatex">\(0.5\)</span> for proteome network. </p>
<p>When the network of islands is constructed in each proteome community, and we found communities in this network we can postulate that each community is a hotspot. From the definition we can see that a hotspot represents a cluster in a network with high density of inner connections, which means sharing similar set of «conserved» flanking protein groups. At this step we add filtering. Firstly, the presence cutoff - if a hotspot, say <span class="arithmatex">\(h_i^c\)</span>, found in proteome community <span class="arithmatex">\(c\)</span> of size <span class="arithmatex">\(N_c\)</span> consists of islands from <span class="arithmatex">\(N_c^{h_i}\)</span> proteomes, then the hotspot is kept only if <span class="arithmatex">\(N_c^{h_i}/N_c \geq 0.3\)</span> (cutoff can be adjusted with <em>hotspot_presence_cutoff</em> config parameter or <code>-hpc, --hotspot-presence-cutoff</code> argument). That means that clusters of variable islands that found only in a small fraction of proteomes within a proteome community are not considered as hotspots.  Additionally, if we have several islands on the same proteome clustered together in one hotspot then the only one with highest sum of weights to other community members will be kept in the hotspot. That can happen if two islands are divided by few (one to three) conserved proteins and this step allows to keep only the one that has higher overlapping rate of flanking proteins with others. That is, the hotspot by our definition can contain only one island per proteome. </p>
<h3 id="hotspot-merging">Hotspot merging</h3>
<p>We can expect that hotspots found in different proteome community will have the same conserved neighbours due to the modularity of mobile elements such as phages or plasmids. In order to detect and process such cases we build the network of hotspots and apply again the Leiden algorithm for community detection. In that case for each hotspot we define attribute <em>signature</em> which represents a set of protein groups that found in at least of <span class="arithmatex">\(0.75\)</span> of its island «conserved» flanking proteins (cutoff can be adjusted with <em>hotspot_signature_presence_cutoff</em> config parameter). That is, <em>hotspot signature</em> is a set of protein groups that found as «conserved» flanking genes in the majority of its islands. Formally, in the case of hotspot network each node, say <span class="arithmatex">\(h_i^c\)</span> represents the corresponding <span class="arithmatex">\(ith\)</span> hotspot of proteome community <span class="arithmatex">\(c\)</span> and each edge, say <span class="arithmatex">\(e_{ij}\)</span>, connecting <span class="arithmatex">\(i^{th}\)</span> and $j^{th} hotspots has weight <span class="arithmatex">\(w_{ij}\)</span>:<br />
$$
w_{ij} = \frac{|G_i \cap G_i|\cdot(|G_i|+|G_j|)}{2|G_i||G_j|}
$$
where <span class="arithmatex">\(|G_i|\)</span>, <span class="arithmatex">\(|G_j|\)</span> are lengths (number of unique elements) of the set of protein groups of <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> hotspot signature, respectively. <span class="arithmatex">\(|G_{i} \cap G_{j}|\)</span> is the number of shared (overlapped) protein groups between <span class="arithmatex">\(i^{th}\)</span> and <span class="arithmatex">\(j^{th}\)</span> hotspot signatures. Network building and community detection parameters set by default are the same as in island network construction and can be adjusted with <em>hotspot_similarity_cutoff</em> and <em>leiden_resolution_parameter_h</em> config parameters.</p>
<h3 id="additional-functional-annotation">Additional functional annotation</h3>
<p>For additional annotation of proteins encoded either as cargo of annotated hotspot islands or as flanking genes we use <a href="https://pyhmmer.readthedocs.io/en/stable/">pyhmmer API</a>. The reason of using pyhmmer instead of MMseqs2 sequence vs profile search is that for several databases that are essential for hotspot protein annotation we have only <em><a href="http://hmmer.org">hmmer</a></em> database of profiles available without provided MSA or HMM models compatible with hh-suite or MMseqs.  In order to reduce running time we use only representative proteins of each protein group as a query set for searching. As it was defined, each cluster of proteins is considered as a set of homologues, then, the search results are attributed to each protein group based on its representative sequence.</p>
<p>The list of used databases and their versions or date of retrieval: </p>
<ul>
<li>Antimicrobial resistance genes (AMR): <a href="https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/AMRFinder/">AMRFinderPlus</a>: 02.05.2024.2</li>
<li>Anti-Defence: <a href="https://bcb.unl.edu/dbAPIS/">dbAPIS_Acr</a>: 19.09.2023</li>
<li>Defence systems: <a href="https://defensefinder.mdmlab.fr">DefenceFinder</a>: 1.2.4, <a href="https://github.com/macsy-models/CasFinder">CasFinder</a>: 3.1.0</li>
<li>Virulence factors: <a href="http://www.mgc.ac.cn/VFs/">VFDB</a> : 10.05.2024</li>
</ul>
<h3 id="lovis4u-visualisation">LoVis4u visualisation</h3>
<p>The last step of the algorithm is using of our tool <a href="https://art-egorov.github.io/lovis4u/">LoVis4u</a> for visualisation of each hotspot community and proteome community. We recommend you to read the <a href="https://art-egorov.github.io/ilund4u/ExampleDrivenGuide/cmd_guide/">Example-driven guide page</a> for examples of visualisation with detailed descriptions. However, there is one step in the visualisation of hotspot communities we want to highlight here: we can expect that some variable islands of one hotspot community from different proteomes could have the same set of cargo and flanking protein groups. In that case iLund4u keeps only the one of them for visualisation in order to avoid duplication on visualisation. Full list of hotspot loci is available in annotation tables.</p>
<p>You can see an example with visualisation of proteome community that consists of 736 proteomes (<a href="../../img/10.pdf" target="_blank"><strong><em>LoVis4u community visualisation pdf</em></strong></a>). The community has one hotspot containing 888 diverse cargo protein groups with 24 of them being annotated as Defence related proteins, 15 as virulence factors, 1 as anti-defence and 1 AMR gene (<a href="../../img/10-0_537-0.pdf" target="_blank"><strong><em>LoVis4u hotspot visualisation pdf</em></strong></a>).</p>
<hr />
<h2 id="ilund4u-database">iLund4u database</h2>
<p>Optionally during the hotspot annotation step you can build the corresponding iLund4u database that holds proteome and hotspot objects and can be used as query for two other modes: protein and proteome annotation. iLund4u database consists of set of files (non-binary) that holds information about protein, proteome, CDS, island and hotspot objects.</p>
<figure style="float: right; margin: 0; padding: 0;">
<a class="glightbox" href="../../img/ilund4u_dbs_wo_header.png" data-type="image" data-width="auto" data-height="auto" data-desc-position="bottom"><img align="right" alt="f2" loading="lazy" margin="0" padding="0" src="../../img/ilund4u_dbs_wo_header.png" width="360" /></a><br />
<figcaption style="width: 100; margin: 0; padding: 0;">Figure 3</figcaption>
</figure>
<p>iLund4u has two precomputed databases of hotspots built on phage and plasmid sequences.<br />
The database of phages was built based on running hotspot annotation mode on all available <a href="https://phagescope.deepomics.org">PhageScope database</a> sequences (~800K sequences, version of June 2024). For plasmids database we took <a href="https://genome.jgi.doe.gov/portal/IMG_PR/IMG_PR.home.html">IMG/PR database of plasmids</a> (~700K sequences, version of June 2024).  </p>
<p><strong>Database sizes:</strong> Phages: 6.2GB; Plasmids: 1.12GB </p>
<div style="clear: both;"></div>

<p>See <a href="https://art-egorov.github.io/ilund4u/ExampleDrivenGuide/cmd_guide/">Example-driven guide page</a> for details about database building and downloading.</p>
<hr />
<h2 id="protein-annotation-mode">Protein annotation mode</h2>
<p>This mode utilises <em>guilt by association</em> (in our case we can say also <em>guilt by location</em>) approach. In short, if an unannotated protein is found in a loci containing several for instance defence systems or in a hotspot which annotated cargo is often a defence systems, then other genes in the locus or in the hotspot are also more likely to be involved in related function. That is, the protein location as well as mobility through hotspots is an important property that can help in functional annotation of proteins with unknown function.   </p>
<h3 id="homology-search">Homology search</h3>
<p>In order to find hotspots containing your query protein homologues we need to define homologues at the first step. iLund4u has two modes for this step. Both starts with <a href="https://mmseqs.com">MMseqs2</a> search of the query protein versus the all proteins of the iLund4u database. By default with the following mmseqs arguments: <code>-e 1e-5 -s 7</code> and subsequent query and target coverage cutoffs set as 0.6. Then we have the differences: </p>
<ol>
<li>In the <em>«group» search mode</em> the query protein is assigned to the same homologous group as that of the best search hit protein.</li>
<li>In the <em>«proteins» search mode</em> all target proteins that pass the search cutoffs are considered to be homologues. </li>
</ol>
<h3 id="hotspots-retrieving">Hotspots retrieving</h3>
<p>If homologues are found in the database, iLund4u searches for hotspots that contain these proteins either as cargo or flanking gene. The search results contain LoVis4u visualisation and set of annotation tables. See <a href="https://art-egorov.github.io/ilund4u/ExampleDrivenGuide/cmd_guide/">Example-driven guide page</a> for details.</p>
<hr />
<h2 id="proteome-annotation-mode">Proteome annotation mode</h2>
<p>The idea of this mode is to annotate hotspot when your query is a single proteome and it has a community of similar proteomes in an iLund4u database. If you have sequenced, let say a phage, then you can search it versus iLund4u database of all phages and in case there is a proteome community of related phages the already known hotspots can be mapped to your query sequence. Additionally it helps do define «conserved» proteins in your query proteome, which homologues are found in the majority of similar proteomes and «variable». See the detailed definition in the corresponding sections above. </p>
<h3 id="homology-search_1">Homology search</h3>
<p>The first step is to define protein group for each CDS of the query proteome. For this task the <em>«group» search mode</em> of <em>protein annotation mode</em> is used. </p>
<h3 id="defining-proteome-community">Defining proteome community</h3>
<p>When protein groups are defined we can calculate proteome similarity weights <span class="arithmatex">\(w_{qi}\)</span> from the query proteome <span class="arithmatex">\(q\)</span> to all other database proteomes. All thresholds for weights are the same as the initial network building described above. </p>
<h3 id="annotation-of-variable-islands-and-hotspot-mapping">Annotation of variable islands and hotspot mapping</h3>
<p>If a community of proteomes was assigned for a query proteome we can continue with defining variable islands. Based on proteome community defining of each protein group class we define query protein group classes values as already defined for community. If we have a new protein group in query proteome it is set as new «variable» protein group for the community. Each variable island then embedded in the network of islands in order to map already annotated hotspots to the annotated islands of the new proteome. </p>
<p>The search results contain LoVis4u visualisation including mapped islands, hotspots, full-length proteome community, as well as several annotation tables. See <a href="https://art-egorov.github.io/ilund4u/ExampleDrivenGuide/cmd_guide/">Example-driven guide page</a> for details.</p>

              
            </article>
            
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../.." class="md-footer__link md-footer__link--prev" aria-label="Previous: Home" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Home
            </div>
          </div>
        </a>
      
      
        
        <a href="../../ExampleDrivenGuide/cmd_guide/" class="md-footer__link md-footer__link--next" aria-label="Next: Example-driven guide" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Example-driven guide
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": ["content.code.annotate", "announce.dismiss", "navigation.tabs", "toc.integrate", "header.autohide"], "search": "../../assets/javascripts/workers/search.b97dbffb.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c99f48ec.min.js"></script>
      
        <script src="../../assets/_markdown_exec_pyodide.js"></script>
      
        <script src="../../javascripts/katex.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/katex.min.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/contrib/auto-render.min.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>